### PR TITLE
ci: add mergify rules for nixpkgs

### PR DESCRIPTION
**Pull request by Mic92**
## Things done

Context: https://github.com/NixOS/org/issues/38

We already deployed [mergify](https://mergify.com/) in nixos-hardware: https://github.com/NixOS/nixos-hardware/pull/1249 and in nix https://github.com/NixOS/nix/blob/master/.mergify.yml
This pull requests adds support for it in nixpkgs. Mergify adds support for declarative merge queues to nixpkgs. Merge queues make sure that pull request are tested against the latest base branch. The advantage over github's own merge queues is that, the configuration is part of the repository and therefore does not require a repository admin to keep "Required Checks" in sync with recently added CI checks.
Another advantage mergify has it's that it is a github app. There are many cases where github won't run checks on automated created pull requests. With mergify this is not an issue. If we are happy with the workflow introduced mergify, we can proceed adding it to `CONTRIBUTING.md`.
We have a similar feature in merge-bot. However it doesn't work very reliable. With mergify, we can later make `merge-bot` just add a label remove the complicated code and database integration we have just now.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Mic92** [comment]
https://github.com/NixOS/org/issues/39

File: .mergify.yml
Comment (JohnRTitor) [inline .mergify.yml:21]: Why rebase? Rebasing gets rid of the signature and recreates the commit with entirely new hash.

Comment (Mic92) [reply .mergify.yml:21]: It makes the git history more readable. I honestly don't see much security win from having those signatures unless we make them mandatory for everyone and also have a way to verify that someones actually owns this ssh key (i.e. keysigning in person). We already see from which account a change came from, if someones github account gets compromised, the attacker might as well just upload a new ssh key. There are a lot of cases in practice already where we amend commits from contributors or rebase on their behalf, so signatures get removed already.
It would be more interesting if we had signatures for merging pull request, but since we use exclusively the "merge" button for that, it will be always the github signing key anyway. Which brings me back to my initial point, we need to trust github and that our users don't loose control over their account.

File: .mergify.yml
Comment (JohnRTitor) [inline .mergify.yml:23]: I suspect this will be used a lot, 5 PRs in queue is just not enough.

File: .mergify.yml
Comment (JohnRTitor) [inline .mergify.yml:20]: By-name and nixfmt checks should not be mandatory, and can be allowed to fail, I believe.

File: .mergify.yml
Comment (Mic92) [reply .mergify.yml:23]: Do you think so? There is a merged pull request every 10 minutes maybe, this would hardly fill up 5 PRs into one batch.

File: .mergify.yml
Comment (Mic92) [reply .mergify.yml:20]: This is for automatic merging, should we not prefer a human looking at the pull request if `nixfmt` or the `by-name` check fails? cc @infinisil 
Afaik `nixfmt` only triggers if the file was formatted before or if it is a new file.

Comment (JohnRTitor) [reply .mergify.yml:20]: Nixfmt is probably fine, but I have seen some cases where nixpkgs-vet fails (like https://github.com/NixOS/nixpkgs/pull/332242)

A human reviewer/commiter is still needed to add the PR to the merge queue.

Comment (Aleksanaa) [reply .mergify.yml:20]: > Nixfmt is probably fine, but I have seen some cases where nixpkgs-vet fails (like #332242)

I believe at least that is a correct one. `firedragon` can and should be put in `by-name`.

File: .mergify.yml
Comment (JohnRTitor) [reply .mergify.yml:23]: Well as commiters most of us just wait for a PR to complete the CI process and then merge. With mergify this process can speed up a lot potentially, multiple reviewers adding multiple PRs to the queue. 


Anyways, I guess we could give it a shot as an initial value, raise it if need arises later.

File: .mergify.yml
Comment (JohnRTitor) [reply .mergify.yml:20]: > I believe at least that is a correct one. `firedragon` can and should be put in `by-name`.

https://github.com/NixOS/nixpkgs/pull/332242#discussion_r1829448213

File: .mergify.yml
Comment (JohnRTitor) [reply .mergify.yml:21]: I used to prefer rebase merge as well, as that would get rid of the merge commit, but moved away from doing that, since most people on Nixpkgs seemed to prefer "preserving the git history" by keeping the old commit hash (no rebase merge).

File: .mergify.yml
Comment (Aleksanaa) [reply .mergify.yml:20]: I'm pretty sure that is a valid case and the error previously reported is not related to the position, but e.g. rebasing to master. The original diff is garbage collected by GitHub so I cannot check it again. When this happens, we can at least check:

1. Can you verify the `outPath` of the package remain the same, when inside `by-name` and outside?
2. Do you have a dynamic reference (generated by function) to files outside the path?

Rather than believe in what people casually says (I have encountered a lot of prevarication replies tbh), we should find out why. If we just randomly throw confusing code into the codebase, we will make it more difficult for those who maintain the code in the future. And if this situation truly reflects a problem with our underlying code, we should at least report it promptly.

File: .mergify.yml
Comment (Mic92) [reply .mergify.yml:23]: Well we can always increase it. This is just a trade off so that if one pull request fails the merge queue doesn't have to split things up. Just now it seems that on average  (using the weekly statistics) we would have one pull request in the pipeline.

File: .mergify.yml
Comment (sersorrel) [reply .mergify.yml:21]: I strongly feel we should avoid rebase merges, not codify them as policy. Whilst not as bad as squash merges, they make it difficult to tell (without referring to GitHub) whether a commit is actually "finished" or "halfway through a PR". Yes, commits should be atomic so it shouldn't matter, but in reality they're not: if, say, a large staging-next PR contains a commit that breaks something, which is then fixed by a followup commit before the merge is done, it's not necessarily the casse that the PR will be modified to combine the breaking and fixing commits; therefore, it is useful to preserve in history "this entire set of changes we merged in was good, but halfway through the PR things might be broken".

Rebase merges also (aiui) make it impossible to use `git branch -d` to delete a branch iff it's been merged, since you lose any machine-readable relationship between "the commits you made" and "the commits that are on `master`".

See also nixpkgs-merge-bot which, afaict, just does "whatever the default is": https://github.com/NixOS/nixpkgs-merge-bot/blob/54faae5c11ae4fc092aabc37c76219fd233da973/nixpkgs_merge_bot/github/GitHubClient.py#L210-L212

Comment (Mic92) [reply .mergify.yml]: Dropped. I understand more and more the appeal of gerrits' changeset id and what jj/sapling is doing. Git simply doesn't have good mechanism for tracking commits that changes after code reviews.

File: .mergify.yml
Comment (JohnRTitor) [reply .mergify.yml:20]: Ok, I have no problem with these tests being a requirement.

**Review by JohnRTitor — APPROVED** [review]
Looks good to me, but I have no way to test this.

**Comment by JohnRTitor** [comment]
Question: Let's say an workflow only runs on certain paths, like https://github.com/NixOS/nixpkgs/pull/361447 (on `lib/*`), if we add that workflow as requirement for mergify, will it work on those PRs where this workflow will not even trigger?

**Comment by JohnRTitor** [comment]
Let's deploy it for now. We can always push a bugfix later, if we need any changes.

**Comment by Mic92** [comment]
I think we would actually mergify installed for this to work. Doesn't think this is the case just now: https://github.com/NixOS/org/issues/39

**Comment by zimbatm** [comment]
Mergify is now active on the repo

**Comment by Mic92** [comment]
Thanks. Looks like merge queues need to be enabled separately.

**Comment by infinisil** [comment]
@Mic92 Enabled now

**Comment by arianvp** [comment]
I just had @Lassulus disable this as it was opening a Pull request **per minute**

https://github.com/NixOS/nixpkgs/pulls?q=is%3Apr+is%3Aclosed+author%3Amergify%5Bbot%5D

**Pull request merged**
